data/Builder.go

```diff
@@ -0,0 +1,110 @@
+package data
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+//Datum represent one data point
+type Datum map[string]interface{}
+
+//Dataset represents an array of data points
+type Dataset []Datum
+
+//GetSelector returns the selector fields for a given platform datatype, or nil if there is no match
+func GetSelector(data interface{}) interface{} {
+
+	switch data.(type) {
+	case *Basal:
+		return data.(*Basal).Selector()
+	case *DeviceEvent:
+		return data.(*DeviceEvent).Selector()
+	default:
+		return nil
+	}
+}
+
+//Builder interface that the TypeBuilder implements
+type Builder interface {
+	BuildFromRaw(raw []byte) (interface{}, *Error)
+	BuildFromData(data map[string]interface{}) (interface{}, *Error)
+	BuildFromDataSet(dataSet Dataset) ([]interface{}, *ErrorSet)
+}
+
+//TypeBuilder that is used to build data types that the platform understands
+type TypeBuilder struct {
+	inject map[string]interface{}
+}
+
+//NewTypeBuilder returns an instance of TypeBuilder
+func NewTypeBuilder(inject map[string]interface{}) Builder {
+	return &TypeBuilder{
+		inject: inject,
+	}
+}
+
+//BuildFromDataSet will build the matching type(s) from the given Dataset
+func (typeBuilder *TypeBuilder) BuildFromDataSet(dataSet Dataset) ([]interface{}, *ErrorSet) {
+
+	var set []interface{}
+	var buildError *ErrorSet
+
+	for i := range dataSet {
+		item, err := typeBuilder.BuildFromData(dataSet[i])
+		if err != nil && !err.IsEmpty() {
+			if buildError == nil {
+				buildError = NewErrorSet()
+			}
+			buildError.AppendError(err)
+			continue
+		}
+		set = append(set, item)
+	}
+	return set, buildError
+}
+
+//BuildFromRaw will build the matching type(s) from the given raw data
+func (typeBuilder *TypeBuilder) BuildFromRaw(raw []byte) (interface{}, *Error) {
```

Comment:
lets go for single letter then. Then we can use the editor to full effect also 

![screen shot 2016-03-24 at 3 22 29 pm](https://cloud.githubusercontent.com/assets/5168993/14006646/564a44d6-f1d4-11e5-8890-a24cc65dbac6.png)